main/broadcasters.c

```diff
@@ -381,23 +384,109 @@ static void mijia_sensor_metadata_get(uint8_t *adv_data, size_t adv_data_len,
+    uint16_t frame_ctrl = le16toh(mijia_header->frame_ctrl);
+    uint8_t *payload_start = (uint8_t *)mijia_header + sizeof(mijia_header_t);
+    uint8_t decrypted[69];
 
     cb("MACAddress", _mactoa(mijia_header->mac), ctx);
     sprintf(s, "%hhu", mijia_header->message_counter);
     cb("MessageCounter", s, ctx);
 
     /* Check if any data is available */
-    if ((be16toh(mijia_header->frame_ctrl) & 0x4000) == 0)
+    if ((frame_ctrl & 0x40) == 0)
         return;
 
-    mijia_data_entry_t *mijia_data_entry = (mijia_data_entry_t *)(
-        (uint8_t *)mijia_header + sizeof(mijia_header_t) +
-        ((be16toh(mijia_header->frame_ctrl) & 0x2000) ? 1 : 0) /* skip capability byte */);
+    /* Check if there's a capability byte */
+    if (frame_ctrl & 0x20)
+    {
+        /* Check if there's an IO capability (uses one more byte) */
+        if (*payload_start & 0x20)
+            payload_start++;
+        payload_start++;
+    }
+
+    /* Check data is encrypted */
+    if (frame_ctrl & 0x08)
+    {
+        mbedtls_ccm_context ctx;
+        uint8_t aad[] = {0x11};
+        uint8_t nonce[12];
+        uint8_t key[16];
+        const char *key_str;
+        uint8_t version = frame_ctrl >> 12;
+        size_t crypt_len;
+        int ret;
+
+        if (version < 4)
+        {
+            ESP_LOGW(TAG, "Legacy MiBeacon encryption not supported, skipping %s", _mactoa(mijia_header->mac));
+            return;
+        }
+
+        key_str = config_ble_mikey_get(_mactoa(mijia_header->mac));
+        if (!key_str)
+        {
+            ESP_LOGE(TAG, "MiBeacon decryption key not found for %s", _mactoa(mijia_header->mac));
+            return;
+        }
+        if (strlen(key_str) != sizeof(key) * 2)
+        {
+            ESP_LOGE(TAG, "MiBeacon decryption key for %s has the wrong length", _mactoa(mijia_header->mac));
+            return;
+        }
+        for (size_t i = 0; i < sizeof(key); i++)
+            sscanf(key_str + 2 * i, "%02hhx", &key[i]);
+
+        mbedtls_ccm_init(&ctx);
+        if (mbedtls_ccm_setkey(&ctx, MBEDTLS_CIPHER_ID_AES, key, 128))
+        {
+            ESP_LOGE(TAG, "Could not set MiBeacon decryption key");
+            return;
+        }
+
+        for (uint8_t i = 0; i < 6; i++)
+            nonce[i] = mijia_header->mac[i];
+        for (uint8_t i = 6; i < 9; i++)
+            nonce[i] = ((uint8_t *)mijia_header)[i - 2];
+        for (uint8_t i = 9; i < 12; i++)
+            nonce[i] = ((uint8_t *)mijia_header)[len - 7 + (i - 9)];
+
+        crypt_len = len - (payload_start - (uint8_t *)mijia_header) - 7;
+        if (crypt_len > sizeof(decrypted))
+        {
+            ESP_LOGE(TAG, "MiBeacon encrypted payload too long: %d", crypt_len);
+            return;
+        }
+
+        ret = mbedtls_ccm_auth_decrypt(&ctx, crypt_len,
+            nonce, sizeof(nonce), aad, sizeof(aad),
+            payload_start, decrypted,
+            (uint8_t *)mijia_header + len - 4, 4);
```

Comment:
I think this can fit a bit nicer:
```c
		ret = mbedtls_ccm_auth_decrypt(&ctx, crypt_len, nonce, sizeof(nonce),
			aad, sizeof(aad), payload_start, decrypted,
			(uint8_t *)mijia_header + len - 4, 4);
```